onnx/defs/shape_inference.h

```diff
@@ -354,6 +354,56 @@ inline void propagateElemTypeFromSequenceInputToOutput(InferenceContext& ctx, si
   }
 }
 
+inline void propagateElemTypeFromOptionalInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
+  auto input_type = ctx.getInputType(inputIndex);
+  if (nullptr == input_type || input_type->value_case() != TypeProto::kOptionalType) {
+    fail_type_inference("Input ", inputIndex, " expected to have optional type");
+  }
+  auto input_opt_type = input_type->optional_type();
+  if (!input_opt_type.has_elem_type()) {
+    fail_type_inference("Element type of optional input ", inputIndex, " unknown");
+  }
+  if (input_opt_type.elem_type().has_tensor_type()) {
+
+	  const auto input_tensor_elem_type = getTensorElementType(input_opt_type.elem_type());
+	  if (input_tensor_elem_type == TensorProto::UNDEFINED) {
+	    fail_type_inference("Tensor type of optional tensor input ", inputIndex, " unknown");
+	  }
+
+	  auto output_opt_type = ctx.getOutputType(outputIndex)->mutable_optional_type();
+	  const auto input_value_case = input_opt_type.elem_type().value_case();
+	  const auto output_value_case = output_opt_type->mutable_elem_type()->value_case();
+	  auto output_tensor_type = output_opt_type->mutable_elem_type();
+	  if (output_value_case == TypeProto::kTensorType || output_value_case == TypeProto::kSparseTensorType) {
+	    setTensorElementType(input_tensor_elem_type, output_value_case, *output_tensor_type);
+	  } else if (output_value_case == TypeProto::VALUE_NOT_SET) {
+	    // Assume output will have the same type
+	    setTensorElementType(input_tensor_elem_type, input_value_case, *output_tensor_type);
+	  } else {
+	    // This is not expected to happen
+	    fail_type_inference(
+	        "Optional tensor output ", outputIndex, " expected to have tensor or sparse tensor type. Got: ", output_value_case);
+	  }
+  } else if (input_opt_type.elem_type().has_sequence_type()) {
```

Comment:
Should there be an 'else' that says propagating non-tensors and non-sequence optionals are unsupported ? (Otherwise the elem_type of the output is undefined/dangling isn't it ? )

Reply:
+1